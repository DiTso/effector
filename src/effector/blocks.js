//@flow

import {step, readRef} from 'effector/stdlib'

//prettier-ignore
export const filterChanged = step.filter({
  fn: (data, {state}) => (
    data !== readRef(state)
    && data !== undefined
  ),
})

export const noop = step.compute({
  fn: n => n,
})

export const updateHandler = step.compute({
  fn: (upd, {handler}) => handler(upd),
})
