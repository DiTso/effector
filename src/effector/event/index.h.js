//@flow
import type {Subscription, Subscriber} from '../index.h'
import type {kind, Graph, Unit} from 'effector/stdlib'
import type {CompositeName} from '../compositeName'

export type Event<E> = /*::interface extends Unit*/ {
  /*::
  [[call]](payload: E): E,
  */
  /*::+*/ id: string,
  getType(): string,
  create(payload: E, type: string, args: any[]): E,
  watch(watcher: (payload: E) => any): Subscription,
  map<T>(fn: (_: E) => T): Event<T>,
  filter<T>(fn: (_: E) => T | void): Event<T>,
  prepend<Before>(fn: (_: Before) => E): Event<Before>,
  subscribe(subscriber: Subscriber<E>): Subscription,
  /*::+*/ kind: kind,
  getType(): string,
  shortName: string,
  graphite: Graph,
  compositeName: CompositeName,
}
