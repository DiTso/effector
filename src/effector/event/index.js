//@flow

export {eventFabric} from './eventFabric'
export {forward} from './forward'
export {createEvent} from './createEvent'
export type {Event} from './index.h'
export {fromObservable} from './fromObservable'
export {filter} from './filter'
