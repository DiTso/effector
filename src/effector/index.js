//@flow

/*
 * People assume that time is a strict progression of cause to effect,
 * but actually from a non-linear, non-subjective viewpoint -
 * it's more like a big ball of wibbly wobbly time-y wimey stuff
 */

export {combine} from './combine'
export {sample} from './sample'

export type {Event} from 'effector/event'
export {createEvent, forward, fromObservable, filter} from 'effector/event'

export type {Effect} from 'effector/effect'
export {createEffect} from 'effector/effect'

export type {Store} from 'effector/store'
export {
  createStore,
  createStoreObject,
  setStoreName,
  createApi,
  restore,
  restoreEvent,
  restoreEffect,
  restoreObject,
} from 'effector/store'

export {Kind, clearNode, createGraph as createNode, step} from 'effector/stdlib'
export {launch} from 'effector/kernel'
export type {kind} from 'effector/stdlib'

export {
  invariant,
  warning,
  isUnit,
  isStore,
  isEvent,
  isEffect,
  isDomain,
  is,
} from 'effector/validate'
export {version} from 'effector/flags'

//eslint-disable-next-line
export * as blocks from 'effector/blocks'
