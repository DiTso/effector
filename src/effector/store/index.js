//@flow

export {storeFabric} from './storeFabric'
export {normalizeConfig} from '../config'
export {createStore} from './createStore'
export {createStoreObject} from './createStoreObject'
export {createApi} from './createApi'
export {restore, restoreEvent, restoreEffect, restoreObject} from './restore'

export type {Store} from './index.h'
