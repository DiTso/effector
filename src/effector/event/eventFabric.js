//@flow
import $$observable from 'symbol-observable'

import {
  step,
  Kind,
  stringRefcount,
  createGraph,
  type Unit,
} from 'effector/stdlib'
import type {Effect} from 'effector/effect'
import {launch} from 'effector/kernel'
import {noop, updateHandler} from 'effector/blocks'

import type {Subscription} from '../index.h'
import type {EventConfigPart} from '../config'
import type {Event} from './index.h'
import {createName} from '../compositeName'
import {forward} from './forward'

const nextID = stringRefcount()

export function eventFabric<Payload>({
  name: nameRaw,
  config = {},
}: {
  name?: string,
  config?: EventConfigPart,
}): Event<Payload> {
  const id = nextID()
  const name = nameRaw || id
  const compositeName = createName(name)
  const fullName = compositeName.fullName
  const graphite = createGraph({
    node: [
      step.emit({
        fullName,
      }),
    ],
  })

  //$off
  const instance: Event<Payload> = (payload: Payload) =>
    instance.create(payload)
  ;(instance: any).getType = () => fullName
  //eslint-disable-next-line no-unused-vars
  ;(instance: any).create = launch.bind(null, instance)
  ;(instance: any).kind = Kind.event
  ;(instance: any)[$$observable] = () => instance
  ;(instance: any).id = id
  ;(instance: any).watch = watchEvent.bind(null, instance)
  ;(instance: any).map = mapEvent.bind(null, instance)
  ;(instance: any).filter = filterEvent.bind(null, instance)
  ;(instance: any).prepend = prepend.bind(null, instance)
  ;(instance: any).subscribe = subscribe.bind(null, instance)
  instance.graphite = graphite
  instance.shortName = name
  instance.compositeName = compositeName
  instance.defaultConfig = config

  return instance
}

const subscribe = (event, observer): Subscription =>
  event.watch(payload => observer.next(payload))

const prepend = (event, fn: (_: any) => *) => {
  const contramapped: Event<any> = eventFabric({})
  forward({
    from: contramapped,
    to: createGraph({
      child: [event],
      scope: {handler: fn},
      node: [updateHandler],
    }),
  })
  return contramapped
}

declare function mapEvent<A, B>(event: Event<A>, fn: (_: A) => B): Event<B>
declare function mapEvent<A, B>(
  effect: Effect<A, any, any>,
  fn: (_: A) => B,
): Event<B>
function mapEvent<A, B>(event: Event<A> | Effect<A, any, any>, fn: A => B) {
  const mapped = eventFabric({})
  forward({
    from: event,
    to: createGraph({
      child: [mapped],
      scope: {handler: fn},
      node: [updateHandler],
    }),
  })
  return mapped
}

function filterEvent<A>(event: Unit, fn: A => boolean): Event<A> {
  const mapped = eventFabric({})
  forward({
    from: event,
    to: createGraph({
      scope: {handler: fn},
      child: [mapped],
      node: [
        step.filter({
          fn: (upd, {handler}) => !!handler(upd),
        }),
      ],
    }),
  })
  return mapped
}

function watchEvent<Payload>(
  event: Unit,
  watcher: (payload: Payload) => any,
): Subscription {
  return forward({
    from: event,
    to: createGraph({
      scope: {handler: watcher},
      //prettier-ignore
      node: [
        noop,
        step.run({
          fn: (payload: Payload, {handler}) => handler(
            payload,
          ),
        }),
      ]
    }),
  })
}
