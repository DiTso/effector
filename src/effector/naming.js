//@flow

import type {CompositeName} from './compositeName'
import {is} from 'effector/validate'
import type {Store} from 'effector/store'
import type {Event} from 'effector/event'
import type {Effect} from 'effector/effect'

export function getDisplayName(unit: {
  compositeName?: CompositeName,
  /*::+*/ id: string,
  /*::...*/
}) {
  if (unit.compositeName) {
    return unit.compositeName.fullName
  }
  return unit.id
}

const unitObjectMaxNames = 25

export function unitObjectName(
  objOrArr:
    | $ReadOnlyArray<Store<any> | Event<any> | Effect<any, any, any> | any>
    | {
        [key: string]: Store<any> | Event<any> | Effect<any, any, any> | any,
      },
) {
  let i = 0
  const arr: Array<any> = Object.values(objOrArr)
  const max = unitObjectMaxNames - 1
  const maxLength = arr.length - 1
  let name = 'combine('
  for (const unit of arr) {
    const comma = i === max || maxLength === i ? '' : ', '
    if (is.store(unit) || is.event(unit) || is.effect(unit)) {
      name += getDisplayName(unit) + comma
    } else {
      name += unit.toString() + comma
    }
    i += 1
    if (comma === '') break
  }
  name += ')'
  return name
}
