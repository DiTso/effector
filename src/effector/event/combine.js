//@flow

import type {Event} from './index.h'
import {eventFabric} from './eventFabric'

export function combine<T>(
  f: (...args: any[]) => T,
  ...events: $ReadOnlyArray<Event<any>>
): Event<T> {
  const unit = eventFabric({})
  const l = events.length
  let awaiting = l
  const values = new Array(l)
  const hasValue = new Array(l).fill(false)
  for (const [index, event] of events.entries()) {
    event.watch(value => {
      if (awaiting > 0) {
        if (!hasValue[index]) {
          hasValue[index] = true
          awaiting -= 1
        }
      }
      values[index] = value
      if (awaiting === 0) {
        unit(f(...values))
      }
    })
  }
  return unit
}
