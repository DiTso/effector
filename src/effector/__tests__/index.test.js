//@flow

import {createEvent, type Event} from 'effector/event'
import {combine} from '../combine'

import {createStoreObject, createStore, type Store} from 'effector/store'
import {show} from 'effector/fixtures/showstep'

test('graphite', () => {
  const fn = jest.fn()
  const fn1 = jest.fn()
  const foo: Event<number> = createEvent('foo')
  const bar = foo.map(x => (fn(x), x + 1))
  const store1: Store<string> = createStore('foo').on(bar, (state, bar) =>
    [state, bar].join(' | '),
  )
  const store2 = store1.map(e => e.length)
  const unsub1 = bar.watch(fn1)
  foo(0)
  foo(10)
  expect(fn).toHaveBeenCalledTimes(2)
  expect(store1.getState()).toBe('foo | 1 | 11')
  // expect(foo.graphite.seq).toMatchSnapshot('foo seq')
  // expect(bar.graphite.seq).toMatchSnapshot('bar seq')
  const showBar = show(bar.graphite)

  expect(showBar).toMatchSnapshot('show bar')
  unsub1()
  const showBar2 = show(bar.graphite)
  expect(showBar2).toMatchSnapshot('show bar')
  // expect(bar.graphite.seq).toMatchSnapshot('bar seq')
  foo(100)
  expect(fn1).toHaveBeenCalledTimes(2)
  // expect(store1.graphite.seq).toMatchSnapshot('store1 seq')
  // expect(store2.graphite.seq).toMatchSnapshot('store2 seq')
  const showFoo = show(foo.graphite)
  const showStore = show(store1.graphite)
  const showStore2 = show(store2.graphite)
  expect(showFoo).toMatchSnapshot('show foo')
  expect(showStore).toMatchSnapshot('show store1')
  expect(showStore2).toMatchSnapshot('show store2')
})

test('showcase', () => {
  const spy = jest.fn()
  const foo = createEvent('foo')
  const bar = createEvent('bar')

  const a = createStore(1)
  const b = createStore(2)
  const bigStore = createStoreObject({a, b})
  const mapped = bigStore.map(s => s.a)

  a.on(foo, n => n + 1)
  b.on(bar, n => n + 1)

  mapped.watch(spy)

  foo()
  foo()

  bar()
  bar()
  bar()
  bar()

  expect(show(a.graphite)).toMatchSnapshot('store a')
  expect(show(foo.graphite)).toMatchSnapshot('event foo')
  expect(show(mapped.graphite)).toMatchSnapshot('mapped')
  expect(spy).toHaveBeenCalledTimes(3)
  const first = createStore('s')
  const second = createStore('h')
  const third = createStore('i')
  const status = combine(first, second, third, (a, b, c) => [a, b, c].join(''))
  expect(status.getState()).toBe('shi')
})
