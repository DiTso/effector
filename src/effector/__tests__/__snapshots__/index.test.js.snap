// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphite: show bar 1`] = `
"Graph {
  Seq [
    1. emit {{
       fullName: 2
    }}
  ]
  Next [
    *  Graph {
       Seq [
         1. compute {{}}
       ]
       Next [
         *  Graph {
            Seq [
              1. filter {{}},
              2. update {{
                 store: {
                   id: 1,
                   current: foo | 1 | 11
                 }
              }},
              3. filter {{}}
            ]
            Next [
              *  Graph {
                 Seq [
                   1. compute {{}},
                   2. filter {{}}
                 ]
                 Next [
                   *  Graph {
                      Seq [
                        1. filter {{}},
                        2. update {{
                           store: {
                             id: 2,
                             current: 12
                           }
                        }},
                        3. filter {{}}
                      ]
                     [(empty list)]}
                ]}
           ]}
      ]},
    *  Graph {
       Seq [
         1. compute {{}},
         2. run {{}}
       ]
      [(empty list)]}
  ]}"
`;

exports[`graphite: show bar 2`] = `
"Graph {
  Seq [
    1. emit {{
       fullName: 2
    }}
  ]
  Next [
    *  Graph {
       Seq [
         1. compute {{}}
       ]
       Next [
         *  Graph {
            Seq [
              1. filter {{}},
              2. update {{
                 store: {
                   id: 1,
                   current: foo | 1 | 11
                 }
              }},
              3. filter {{}}
            ]
            Next [
              *  Graph {
                 Seq [
                   1. compute {{}},
                   2. filter {{}}
                 ]
                 Next [
                   *  Graph {
                      Seq [
                        1. filter {{}},
                        2. update {{
                           store: {
                             id: 2,
                             current: 12
                           }
                        }},
                        3. filter {{}}
                      ]
                     [(empty list)]}
                ]}
           ]}
      ]}
  ]}"
`;

exports[`graphite: show foo 1`] = `
"Graph {
  Seq [
    1. emit {{
       fullName: foo
    }}
  ]
  Next [
    *  Graph {
       Seq [
         1. compute {{}}
       ]
       Next [
         *  Graph {
            Seq [
              1. emit {{
                 fullName: 2
              }}
            ]
            Next [
              *  Graph {
                 Seq [
                   1. compute {{}}
                 ]
                 Next [
                   *  Graph {
                      Seq [
                        1. filter {{}},
                        2. update {{
                           store: {
                             id: 1,
                             current: foo | 1 | 11 | 101
                           }
                        }},
                        3. filter {{}}
                      ]
                      Next [
                        *  Graph {
                           Seq [
                             1. compute {{}},
                             2. filter {{}}
                           ]
                           Next [
                             *  Graph {
                                Seq [
                                  1. filter {{}},
                                  2. update {{
                                     store: {
                                       id: 2,
                                       current: 18
                                     }
                                  }},
                                  3. filter {{}}
                                ]
                               [(empty list)]}
                          ]}
                     ]}
                ]}
           ]}
      ]}
  ]}"
`;

exports[`graphite: show store1 1`] = `
"Graph {
  Seq [
    1. filter {{}},
    2. update {{
       store: {
         id: 1,
         current: foo | 1 | 11 | 101
       }
    }},
    3. filter {{}}
  ]
  Next [
    *  Graph {
       Seq [
         1. compute {{}},
         2. filter {{}}
       ]
       Next [
         *  Graph {
            Seq [
              1. filter {{}},
              2. update {{
                 store: {
                   id: 2,
                   current: 18
                 }
              }},
              3. filter {{}}
            ]
           [(empty list)]}
      ]}
  ]}"
`;

exports[`graphite: show store2 1`] = `
"Graph {
  Seq [
    1. filter {{}},
    2. update {{
       store: {
         id: 2,
         current: 18
       }
    }},
    3. filter {{}}
  ]
  [(empty list)]}"
`;

exports[`showcase: event foo 1`] = `
"Graph {
  Seq [
    1. emit {{
       fullName: foo
    }}
  ]
  Next [
    *  Graph {
       Seq [
         1. compute {{}}
       ]
       Next [
         *  Graph {
            Seq [
              1. filter {{}},
              2. update {{
                 store: {
                   id: 3,
                   current: 3
                 }
              }},
              3. filter {{}}
            ]
            Next [
              *  Graph {
                 Seq [
                   1. filter {{}},
                   2. tap {{}},
                   3. tap {{}},
                   4. barrier {{
                      barrierID: 1
                   }},
                   5. compute {{}}
                 ]
                 Next [
                   *  Graph {
                      Seq [
                        1. filter {{}},
                        2. update {{
                           store: {
                             id: 5,
                             current: {
                               a: 3,
                               b: 6
                             }
                           }
                        }},
                        3. filter {{}}
                      ]
                      Next [
                        *  Graph {
                           Seq [
                             1. compute {{}},
                             2. filter {{}}
                           ]
                           Next [
                             *  Graph {
                                Seq [
                                  1. filter {{}},
                                  2. update {{
                                     store: {
                                       id: 7,
                                       current: 3
                                     }
                                  }},
                                  3. filter {{}}
                                ]
                                Next [
                                  *  Graph {
                                     Seq [
                                       1. compute {{}},
                                       2. run {{}}
                                     ]
                                    [(empty list)]}
                               ]}
                          ]}
                     ]}
                ]}
           ]}
      ]}
  ]}"
`;

exports[`showcase: mapped 1`] = `
"Graph {
  Seq [
    1. filter {{}},
    2. update {{
       store: {
         id: 7,
         current: 3
       }
    }},
    3. filter {{}}
  ]
  Next [
    *  Graph {
       Seq [
         1. compute {{}},
         2. run {{}}
       ]
      [(empty list)]}
  ]}"
`;

exports[`showcase: store a 1`] = `
"Graph {
  Seq [
    1. filter {{}},
    2. update {{
       store: {
         id: 3,
         current: 3
       }
    }},
    3. filter {{}}
  ]
  Next [
    *  Graph {
       Seq [
         1. filter {{}},
         2. tap {{}},
         3. tap {{}},
         4. barrier {{
            barrierID: 1
         }},
         5. compute {{}}
       ]
       Next [
         *  Graph {
            Seq [
              1. filter {{}},
              2. update {{
                 store: {
                   id: 5,
                   current: {
                     a: 3,
                     b: 6
                   }
                 }
              }},
              3. filter {{}}
            ]
            Next [
              *  Graph {
                 Seq [
                   1. compute {{}},
                   2. filter {{}}
                 ]
                 Next [
                   *  Graph {
                      Seq [
                        1. filter {{}},
                        2. update {{
                           store: {
                             id: 7,
                             current: 3
                           }
                        }},
                        3. filter {{}}
                      ]
                      Next [
                        *  Graph {
                           Seq [
                             1. compute {{}},
                             2. run {{}}
                           ]
                          [(empty list)]}
                     ]}
                ]}
           ]}
      ]}
  ]}"
`;
